src/core/tiledscene/qgstiledscenelayerrenderer.cpp

```diff
@@ -75,51 +90,550 @@ bool QgsTiledSceneLayerRenderer::render()
+  const double mapYCenter = 0.5 * ( mapExtent.yMinimum() + mapExtent.yMaximum() );
+  const double mapWidthMeters = context->distanceArea().measureLine(
+                                  QgsPointXY( mapExtent.xMinimum(), mapYCenter ),
+                                  QgsPointXY( mapExtent.xMaximum(), mapYCenter )
+                                );
+  const double mapMetersPerPixel = mapWidthMeters / context->outputSize().width();
+  const double maximumErrorInMeters = maximumErrorPixels * mapMetersPerPixel;
+
+  QgsTiledSceneRequest request;
+  request.setFeedback( feedback() );
+
+  // TODO what z range makes sense here??
+  const QVector< QgsVector3D > corners = QgsBox3D( mapExtent, -1000, 1000 ).corners();
+  QVector< double > x;
+  x.reserve( 8 );
+  QVector< double > y;
+  y.reserve( 8 );
+  QVector< double > z;
+  z.reserve( 8 );
+  for ( int i = 0; i < 8; ++i )
+  {
+    const QgsVector3D &corner = corners[i];
+    x.append( corner.x() );
+    y.append( corner.y() );
+    z.append( corner.z() );
+  }
+  mSceneToMapTransform.transformInPlace( x, y, z, Qgis::TransformDirection::Reverse );
+
+  const auto minMaxX = std::minmax_element( x.constBegin(), x.constEnd() );
+  const auto minMaxY = std::minmax_element( y.constBegin(), y.constEnd() );
+  const auto minMaxZ = std::minmax_element( z.constBegin(), z.constEnd() );
+  request.setFilterBox(
+    QgsOrientedBox3D::fromBox3D( QgsBox3D( *minMaxX.first, *minMaxY.first, *minMaxZ.first, *minMaxX.second, *minMaxY.second, *minMaxZ.second ) )
+  );
 
-  bool canceled = false;
+  request.setRequiredGeometricError( maximumErrorInMeters );
 
-  if ( mLayerBoundingVolume )
+  return request;
+}
+
+bool QgsTiledSceneLayerRenderer::renderTiles( QgsTiledSceneRenderContext &context )
+{
+  const QgsRectangle mapExtent = context.renderContext().mapExtent();
+  auto tileIsVisibleInMap = [mapExtent, this]( const QgsTiledSceneTile & tile )->bool
+  {
+    // the trip from map CRS to scene CRS will have expanded out the bounding volumes for the tile request, so
+    // we want to cull any tiles which we've been given which don't actually intersect our visible map extent
+    // when we transform them back into the destination map CRS.
+    // This potentially saves us requesting data for tiles which aren't actually visible in the map.
+    const QgsGeometry tileGeometry( tile.boundingVolume().as2DGeometry( mSceneToMapTransform ) );
+    return tileGeometry.intersects( mapExtent );
+  };
+
+  QgsTiledSceneRequest request = createBaseRequest();
+  QVector< long long > tileIds = mIndex.getTiles( request );
+  while ( !tileIds.empty() )
+  {
+    if ( feedback() && feedback()->isCanceled() )
+      return false;
+
+    const long long tileId = tileIds.first();
+    tileIds.pop_front();
+
+    const QgsTiledSceneTile tile = mIndex.getTile( tileId );
+    if ( !tile.isValid() || !tileIsVisibleInMap( tile ) )
+      continue;
+
+    switch ( mIndex.childAvailability( tileId ) )
```

Comment:
is all this necessary? If the request did not have the `NoHierarchyFetch` flag set, I would have expected that all relevant children should be available?

Reply:
It's designed this way so that we can apply the extra short cut logic from https://github.com/qgis/QGIS/blob/2a913c4efa77babd8c2ef2101c218a7057212c65/src/core/tiledscene/qgstiledscenelayerrenderer.cpp#L159 .

Otherwise we end up requesting data from outside the visible map extent in a lot of cases (verified!).